uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -0,0 +1,642 @@
+//! i.e. it takes ownership of the `RustFuture` _via_ `Box`!
+//!
+//! Similarly to the _poll function_, it forwards everything to
+//! [`uniffi_rustfuture_drop`], which is the generic version of the monomorphised _drop
+//! function_.
+//!
+//! ## How does `Future` work exactly?
+//!
+//! A [`Future`] in Rust does nothing. When calling an async function, it just
+//! returns a `Future` but nothing has happened yet. To start the computation,
+//! the future must be polled. It returns [`Poll::Ready(r)`][`Poll::Ready`] if
+//! the result is ready, [`Poll::Pending`] otherwise. `Poll::Pending` basically
+//! means:
+//!
+//! > Please, try to poll me later, maybe the result will be ready!
+//!
+//! This model is very different than what other languages do, but it can actually
+//! be translated quite easily, fortunately for us!
+//!
+//! But… wait a minute… who is responsible to poll the `Future` if a `Future` does
+//! nothing? Well, it's _the executor_. The executor is responsible _to drive_ the
+//! `Future`: that's where they are polled.
+//!
+//! But… wait another minute… how does the executor know when to poll a [`Future`]?
+//! Does it poll them randomly in an endless loop? Well, no, actually it depends
+//! on the executor! A well-designed `Future` and executor work as follows.
+//! Normally, when [`Future::poll`] is called, a [`Context`] argument is
+//! passed to it. It contains a [`Waker`]. The [`Waker`] is built on top of a
+//! [`RawWaker`] which implements whatever is necessary. Usually, a waker will
+//! signal the executor to poll a particular `Future`. A `Future` will clone
+//! or pass-by-ref the waker to somewhere, as a callback, a completion, a
+//! function, or anything, to the system that is responsible to notify when a
+//! task is completed. So, to recap, the waker is _not_ responsible for waking the
+//! `Future`, it _is_ responsible for _signaling_ the executor that a particular
+//! `Future` should be polled again. That's why the documentation of
+//! [`Poll::Pending`] specifies:
+//!
+//! > When a function returns `Pending`, the function must also ensure that the
+//! > current task is scheduled to be awoken when progress can be made.
+//!
+//! “awakening” is done by using the `Waker`.
+//!
+//! ## Awaken from the foreign language land
+//!
+//! Our _poll function_ receives a waker function pointer, along with a waker
+//! environment. We said that the waker function lives on the foreign language
+//! side. That's really important. It cannot live inside Rust because Rust
+//! isn't aware of which foreign language it is run from, and thus doesn't know
+//! which executor is used. It is UniFFI's job to write a proper foreign waker
+//! function that will use the native foreign language's executor provided
+//! by the foreign language itself (e.g. `Task` in Swift) or by some common
+//! libraries (e.g. `asyncio` in Python), to ask to poll the future again.
```

Comment:
In theory the foreign side could send a different waker for each poll.  Can you document what would happen in that case and if this ever actually happens in practice.

Also, I think there's an implied lifetime here: The waker can't be freed before the drop function is called.  Can you write some words on that?

Reply:
I've added some words about it. Having a different waker per future is technically possible but I've decided to say it will lead to undefined behaviours. Same for the waker's lifetime, it must be greater than the `RustFuture`'s lifetime, otherwise it's definitely UB.